File: 46.全排列.js
/*
 * @lc app=leetcode.cn id=46 lang=javascript
 *
 * [46] 全排列
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums2) {
	const arr = []
	function fn(nums, ret = []) {
		if (ret.length === nums2.length) {
			arr.push([...ret])
		}
		for (const n in nums) {
			ret.push(nums[n])
			const tp = [...nums]
			tp.splice(n, 1)
			fn(tp, ret)
			ret.pop()
		}
	}
	fn(nums2)
	return arr
}
// @lc code=end

File: 46.全排列.go
/*
 * @lc app=leetcode.cn id=46 lang=golang
 *
 * [46] 全排列
 */

// @lc code=start
func permute(nums []int) [][]int {
	size := len(nums)
	arr := make([][]int, 0)
	r := make([]int, 0)
	used := make([]bool, size)
	var fn func(nums []int, ret []int)
	fn = func(nums []int, ret []int) {
		if len(ret) == size {
			r := make([]int, size)
			copy(r, ret)
			arr = append(arr, r)
		}
		for i, num := range nums {
			if !used[i] {
				used[i] = true
				ret = append(ret, num)
				fn(nums, ret)
				ret = ret[:len(ret)-1]
				used[i] = false
			}
		}
	}
	fn(nums, r)
	return arr
}
// @lc code=end

